src/qibo/backends/numpy.py

```diff
@@ -105,24 +106,28 @@ def matrix(self, gate):
         """Convert a gate to its matrix representation in the computational basis."""
         name = gate.__class__.__name__
         _matrix = getattr(self.matrices, name)
-        return _matrix(2 ** len(gate.target_qubits)) if callable(_matrix) else _matrix
+        if callable(_matrix):
+            _matrix = _matrix(2 ** len(gate.target_qubits))
+
+        return self.cast(_matrix, dtype=_matrix.dtype)
 
     def matrix_parametrized(self, gate):
         """Convert a parametrized gate to its matrix representation in the computational basis."""
         name = gate.__class__.__name__
-        return getattr(self.matrices, name)(*gate.parameters)
+        matrix = getattr(self.matrices, name)(*gate.parameters)
+        return self.cast(matrix, dtype=matrix.dtype)
 
     def matrix_fused(self, fgate):
         rank = len(fgate.target_qubits)
-        matrix = np.eye(2**rank, dtype=self.dtype)
+        matrix = np.eye(2**rank)
```

Comment:
Since the matrices are generally small, the entire fusion of gates is done with `numpy` arrays and then cast to the corresponding backend. That is faster --- especially for GPU backends --- than casting all the matrices to the corresponding backend during calculation.